scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java

```diff
@@ -171,8 +206,8 @@ private void simulateSchedule(Map<ActivityInstanceId, Pair<Duration, SerializedA
    * @param activityInstanceId the activity id
    * @return its duration if the activity has been simulated and has finished simulating, an IllegalArgumentException otherwise
    */
-  public Duration getActivityDuration(ActivityInstanceId activityInstanceId){
-    return engine.getTaskDuration(plannedDirectiveToTask.get(activityInstanceId));
+  public Optional<Duration> getActivityDuration(ActivityInstanceId activityInstanceId){
+    return Optional.of(engine.getTaskDuration(plannedDirectiveToTask.get(activityInstanceId)));
```

Comment:
>Second commit is making the duration getters return Optional so it is cleaner from the outside.

Looks like an optional is unnecessary here. Maybe I'm missing some context?

Reply:
Ahh, it looks like you've made lower-level functions return `Optional` to make it easier on higher-level functions. I think any `Optional` wrapping should be done client-side (within the higher-level code), unless `engine.getTaskDuration()` returned a `null` or `Optional` itself.

Reply:
So actually, my code is wrong because the SimulationEngine throws an exception when the duration is not available. I should try/catch it. And what you are saying is that it should be passed up to the simulation facade instead of treated in the driver ?

Reply:
Ohh, good catch, I missed that.

```java
  public Duration getTaskDuration(TaskId taskId){
    final var state = tasks.get(taskId);
    if (state instanceof ExecutionState.Terminated e) {
      return e.joinOffset().minus(e.startOffset());
    }
    throw new IllegalStateException("Asking for the duration of unfinished task");
  }
```

It may be a good idea to change this exception to a custom checked exception (instead of an unchecked runtime exception like it is now). Then you can choose to either bubble up the exception or do a try/catch within `IncrementalSimulationDriver` to wrap it as an `Optional` if the exception is thrown.

All depends on whether the incremental sim. driver is responsible for gracefully handling the exception case or not. I have a feeling that the incremental sim. driver should bubble the exception up and have clients (callers) worry about how it should be handled (`Optional` wrapping, logging an error, etc.).

Reply:
Hmm... I'm not arguing either way, but I do think unconditionally wrapping a value in an Optional is suspicious.

I propose having the simulation engine return an optional.

Just so I understand - under what circumstances would we call getActivityDuration and expect it to be empty?

Reply:
@mattdailis this can happen when the simulation of an activity has started but has not finished and you are asking for its duration.

Reply:
Querying the duration of an unfinished task is not a normal operation/a normal interaction to have with the engine, it denotes an algorithmic failure on the client side, that's why I put a runtime exception. What do you think about that ?

Reply:
So it's not expected that any client (UI or otherwise) will ask for an unfinished task's duration, right? If so then I agree, the `IllegalStateException` runtime ex. sounds okay. But if there's any UI/client-side workflow that regularly queries for an unfinished task's duration then a checked exception would probably be best.

Reply:
It is not expected no. I'll let the exception in the engine and remove the Optional in the client.

Reply:
Actually... querying a duration of an unfinished task is kind of the only way of knowing whether a task has finished or not. The other way is computing `SimulationResults` but this is very costly. 
I retract my previous statement and I am going with Matt's proposition, the engine returning an `Optional` directly.  
Sorry for the back and forth.

Reply:
Looks good to me